flag_value_source.go

```diff
@@ -0,0 +1,56 @@
+package cli
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"syscall"
+)
+
+// FlagValueSource encapsulates a source which can be used to
+// fetch a value
+type FlagValueSource interface {
+	// Returns the value from the source and if it was found
+	// otherwise returns an empty string & found is set to false
+	Get() (string, bool)
+
+	// The identifier for this source
+	Identifier() string
+}
+
+// ValueSources encapsulates all value sources
+type ValueSources []FlagValueSource
+
+func (v ValueSources) Get() (string, string, bool) {
+	for _, src := range v {
+		if value, found := src.Get(); found {
+			return value, src.Identifier(), true
+		}
+	}
+
+	return "", "", false
+}
+
+// EnvSource encapsulates an env
+type EnvSource string
+
+func (e EnvSource) Get() (string, bool) {
+	envVar := strings.TrimSpace(string(e))
+	return syscall.Getenv(envVar)
+}
+
+func (e EnvSource) Identifier() string {
+	return fmt.Sprintf("environment variable %q", string(e))
+}
+
+// FileSource encapsulates an file source
+type FileSource string
+
+func (f FileSource) Get() (string, bool) {
```

Comment:
Not returning an error could hide important information. For example wrong permissions would cause this source to silently be ignored.
I see that this behavior is consistent with `flagFromEnvOrFile`, but maybe it can be improved.

Reply:
So far no one has asked for it but yes I think it might be worthwhile.

Reply:
The problem with err is should it be ignored or not. Some may want to ignore and continue to next source, others may want to return an error to the cli user. So that means we have to setup a policy on whether to ignore those errors or not. Add in flag validation as well and it starts getting complicated. So for now lets stick to current behaviour and then re-evaulate when the feature request actually comes up